### PR TITLE
(multi trie management) perf: make binary nodes compute their edges in parallel

### PR DESCRIPTION
**Pull request by cchudant**
Hello, this PR is on top of `add_multi_trie_management` for now.

During the commit of a MerkleTree, when we encounter a Binary node, we can compute the hash of both of the edges in parallel.
Hash computation and committing have been separated, the former one being `rayon`ized and pushing to a vec, which the latter one consumes

Feel free to ask questions around the `rayon::join` part, it's a low level rayon primitive made for this kind of case but it can be a bit obscure 

@AurelienFT

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -168,6 +168,19 @@ where
     tries: MerkleTrees<H, DB, ChangeID>,
 }
 
+impl<ChangeID, DB, H> Clone for BonsaiStorage<ChangeID, DB, H>
+where
+    DB: BonsaiDatabase + Clone,
+    ChangeID: id::Id,
+    H: StarkHash + Send + Sync,
+{
+    fn clone(&self) -> Self {
```
Comment (AurelienFT) [inline src/lib.rs:178]: Why did you add the clone implementation ? It seems really heavy to clone and I don't this the use case.

File: src/key_value_db.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@ use crate::{
 };
 
 /// Crate Trie <= KeyValueDB => BonsaiDatabase
+#[derive(Clone)]
```
Comment (AurelienFT) [inline src/key_value_db.rs]: How can you clone this structure if DB isn't Cloneable ? I don't really the fact we allow this. Can you explain why ?

File: src/lib.rs (hunk 1 above)
Comment (cchudant) [reply src/lib.rs:178]: I don't like it either.. but I need to `Clone` the bonsai-trie with a whole snapshot of the hashmapdb in the bench

Maybe #[cfg_attr(test, derive(Clone)]..?

The clone is done here https://github.com/massalabs/bonsai-trie/pull/4/files/5ac51f0ce2b9ddbae42b0eab1e6ae37c53876c84#diff-d9e80af4d48cbcefd1396330b5e5ecbfef4de8fa4882f85098e2c0e3baa9319fR42

Comment (AurelienFT) [reply src/lib.rs:178]: Yes can you hide the clone behind a feature and you use this features in your benches ?

Comment (cchudant) [reply src/lib.rs:178]: benches compile with feature test iirc so #[cfg_attr(test, derive(Clone)] works i think (i'll update the pr)

**Comment by cchudant** [comment]
@AurelienFT I have addressed the clone issue

We will most likely ship this in deoxys right now, although having this many forks all over the place is messy
will this get into multi trie management, or should I transfer this to the main repo once multi trie is merged? (when will it get merged, by the way? - im totally out of the loop) 

I also tried optimizing a little bit: get a bit lower level with a rayon::join_context in a way that allows to reuse the destination array without having to make a new local one if the compute job for the right path did not get stolen
unfortunately it's unclear whether it improved perf or not in the benches, so it's not included in the pr

it might because the machine I am currently using for benching has high variation in the results, i might reexplore this at some point
